### PR TITLE
Sanchonet - Support for new certificate types 

### PR DESCRIPTION
**Pull request by Sotatek-HuyLe3a**

### CODE REVIEW

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/domain/CommitteeRegistration.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+@AllArgsConstructor
+@SuperBuilder
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+public class CommitteeRegistration extends BlockAwareDomain {
+    private String txHash;
+
+    private long certIndex;
+
+    private Long slot;
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/domain/CommitteeRegistration.java:23]: Please also add epoch to every domain model/entity, as the epoch is a key attribute for the ratification and enactment of governance proposals.

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/domain/DRepRegistration.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    private long certIndex;
+
+    private CertificateType type;
+
+    private Long slot;
+
+    private BigInteger deposit;
+
+    private String drepHash;
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/domain/DRepRegistration.java:33]: As mentioned before, please add credential type.

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/domain/DelegationVote.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    private String txHash;
+
+    private long certIndex;
+
+    private Long slot;
+
+    private String address;
+
+    private String drepHash;
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/domain/DelegationVote.java:28]: Same as above

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/processor/CommitteeRegistrationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                            .txHash(txHash)
+                            .slot(eventMetadata.getSlot())
+                            .blockNumber(eventMetadata.getBlock())
+                            .blockTime(eventMetadata.getBlockTime())
+                            .coldKey(resignCommitteeColdCert.getCommitteeColdCredential().getHash())
+                            .certIndex(index)
+                            .build();
+
+                    if (resignCommitteeColdCert.getAnchor() != null) {
+                        committeeDeRegistration.setAnchorHash(resignCommitteeColdCert.getAnchor().getAnchor_data_hash());
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/processor/CommitteeRegistrationProcessor.java:56]: Looks like we need to rename anchor_data_hash field to anchorDataHash in Yaci.

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/processor/DRepRegistrationProcessor.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                DRepRegistration drepRegistration = switch (certificate.getType()) {
+                    case REG_DREP_CERT -> {
+                        RegDrepCert regDrepCert = (RegDrepCert) certificate;
+                        yield buildDRepRegistration(regDrepCert.getDrepCredential(), regDrepCert.getAnchor(),
+                                regDrepCert.getCoin(), certificate.getType(),
+                                txHash, index, eventMetadata);
+                    }
+                    case UNREG_DREP_CERT -> {
+                        UnregDrepCert unregDrepCert = (UnregDrepCert) certificate;
+                        yield buildDRepRegistration(unregDrepCert.getDrepCredential(), null, unregDrepCert.getCoin().negate(),
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/processor/DRepRegistrationProcessor.java:52]: Do we need to negate this? I think we can store positive value or whatever is there in the certificate.
From certificate type, we can decide that it's for refund.

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/processor/DRepRegistrationProcessor.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                                                   int certIndex, EventMetadata eventMetadata) {
+        DRepRegistration drepRegistration = DRepRegistration.builder()
+                .txHash(txHash)
+                .certIndex(certIndex)
+                .type(certificateType)
+                .slot(eventMetadata.getSlot())
+                .blockNumber(eventMetadata.getBlock())
+                .blockTime(eventMetadata.getBlockTime())
+                .drepHash(drepCredential.getHash())
+                .drepId(drepCredential.getHash() != null ? DRepId.fromKeyHash(drepCredential.getHash()) : null)
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/processor/DRepRegistrationProcessor.java:85]: @Sotatek-HuyLe3a Can you please add a //TODO here ? We need to check if the DRepId is different for script hash. Otherwise, we have to change the method name from  DRepId.fromKeyHash to DRepId.fromCredential

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/storage/impl/model/CommitteeDeRegistrationEntity.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@SuperBuilder
+@Entity
+@Table(name = "committee_de_registration")
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/storage/impl/model/CommitteeDeRegistrationEntity.java:15]: Let's change the table name to committee_deregistration

File: stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/util/AddressUtil.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+import lombok.NonNull;
+
+import static com.bloxbean.cardano.client.address.util.AddressEncoderDecoderUtil.*;
+
+//TODO -- Remove this class once the cardano-client-lib is updated
+public class AddressUtil {
+    private static final byte RWD_STAKE_KEY_HEADER_TYPE = (byte)0b1110_0000;
+    private static final byte RWD_STAKE_SCRIPT_HEADER_TYPE = (byte)0b1111_0000;
+
+    public static Address getRewardAddress(@NonNull StakeCredential stakeCredential, boolean isMainnet) {
```
Comment (satran004) [inline stores/governance/src/main/java/com/bloxbean/cardano/yaci/store/governance/util/AddressUtil.java:20]: This method is now there in CCL 0.5.0.  I think we don't need this class any more.

```
Address getRewardAddress(@NonNull Credential stakeCredential, Network networkInfo)
```

File: stores/governance/src/main/resources/db/store/postgresql/V0_1100_1__init.sql [hunk 9]
```diff
@@ -41,6 +41,64 @@ create table voting_procedure
...
+    anchor_hash     varchar(64),
+    cold_key        varchar     NOT NULL,
+    slot            bigint,
+    block           bigint,
+    block_time      bigint,
+    update_datetime timestamp,
+    PRIMARY KEY (tx_hash, cert_index)
+);
+
+CREATE TABLE delegation_vote
```
Comment (satran004) [inline stores/governance/src/main/resources/db/store/postgresql/V0_1100_1__init.sql:71]: We may need additional index for find delegation for a stake address etc. But let's do it in another PR.

**Review by satran004 — COMMENTED** [review]
Thanks @Sotatek-HuyLe3a. Overall looks good to me. Few comments. Please check

File: stores-api/governance-api/src/main/java/com/bloxbean/cardano/yaci/store/api/governance/controller/CommitteeController.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+@RequiredArgsConstructor
+@Tag(name = "Committee Service")
+@RequestMapping("${apiPrefix}/committees")
+@ConditionalOnExpression("${store.governance.endpoints.committee.enabled:true}")
+public class CommitteeController {
+    private final CommitteeRegistrationService committeeRegistrationService;
+    private final CommitteeDeRegistrationService committeeDeRegistrationService;
+
+    @GetMapping("/registrations")
+    @Operation(description = "Get committee registrations by page number and count")
```
[the remark this answers is not in the document]
Comment (satran004) [reply stores-api/governance-api/src/main/java/com/bloxbean/cardano/yaci/store/api/governance/controller/CommitteeController.java:33]: I noticed that queries are in descending order of slot in storage impl. It should be enough for now.
You can also create a separate PR to handle order by parameter if it requires too much change.

**Comment by Sotatek-HuyLe3a** [comment]
@satran004 , I updated code after your first code review, please review again. Thanks.

**Review by satran004 — APPROVED** [review]
Thanks @Sotatek-HuyLe3a
